Fixes 1872

Olink opened the pull request:
In order to track fluid bombs/rockets, we now track all recently created projectiles that we track for fluid creation.  If a bomb exists near where fluid is being created, we allow it.

Fixes #1872
Fixes #1884
Fixes #1885
Fixes #1886
Fixes #1888
Fixes #1889
Fixes #1893

Olink committed:
Add a projectile tracker, so that we can allow fluid bombs.

Olink committed:
Add entry to changelog.

Olink committed:
Make bouncer compile.  Reduce range checks.

hakusaro commented on TShockAPI/Bouncer.cs:
```diff
@@ -1290,6 +1302,18 @@ internal void OnLiquidSet(object sender, GetDataHandlers.LiquidSetEventArgs args
 			}
 
+			bool wasThereABombNearby = false;
+
```
```suggestion
// The +/- 16 is compensation for maximum detonation distance, but it might be possible to drop to 5 and be conservative
```

hakusaro commented on CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ This is the rolling changelog for TShock for Terraria. Use past tense when addin
 * Update for OTAPI 2.0.0.36 and Terraria 1.4.0.4. (@hakusaro, @Patrikkk, @DeathCradle)
 * Fixed /wind command. (@AxeelAnder)
 * Fixed NPC buff bouncer. (@AxeelAnder)
+* Fix fluid bombs. (@Olink)
```
```suggestion
* Fixed lava, wet, and honey bombs. (@Olink)
```

hakusaro commented on TShockAPI/TSPlayer.cs:
```diff
@@ -764,6 +764,12 @@ public bool HasModifiedIceSuccessfully(int x, int y, short tileType, GetDataHand
 		public int LastKilledProjectile = 0;
 
+		/// <summary>
+		/// Keeps track of recently created projectiles by this player.  TShock.cs OnSecondUpdate() removes from this in an async task.
```
```suggestion
		/// Keeps track of recently created projectiles by this player. TShock.cs OnSecondUpdate() removes from this in an async task.
```

hakusaro commented on TShockAPI/TSPlayer.cs:
```diff
@@ -764,6 +764,12 @@ public bool HasModifiedIceSuccessfully(int x, int y, short tileType, GetDataHand
 
+		/// <summary>
+		/// Keeps track of recently created projectiles by this player.  TShock.cs OnSecondUpdate() removes from this in an async task.
+		/// Projectiles older than 5 seconds are purged from this collection as they are no longer "recent".
```
```suggestion
		/// Projectiles older than 5 seconds are purged from this collection as they are no longer "recent."
```

hakusaro commented on TShockAPI/TShock.cs:
```diff
@@ -1064,6 +1065,18 @@ private void OnSecondUpdate()
 					}
 				}
+
+				Task.Run(() =>
```
If you have any other better ideas for this location (bouncer?) I would prefer that there more than here.

Olink replied:
Will do.

hakusaro commented on TShockAPI/GetDataHandlers.cs:
```diff
@@ -2386,6 +2386,10 @@ private static bool HandleProjectileKill(GetDataHandlerArgs args)
 			}
 
 			args.Player.LastKilledProjectile = type;
+			lock (args.Player.RecentlyCreatedProjectiles)
```
Should this be in bouncer or do we want this to be here? No need to move atm.

Olink replied:
Due to the nature of the property, tracking recent projectiles, we may want to keep it here for now.  Ideally, none of our checks should be in GetDataHandlers but for now thats where most of our code lives.

hakusaro commented on TShockAPI/GetDataHandlers.cs:
```diff
@@ -3446,6 +3450,13 @@ public enum EditType
 			{ ProjectileID.MysticSnakeCoil, TileID.MysticSnakeRope }
 		};
 
+		internal static Dictionary<int, LiquidType> projectileCreatesLiquid = new Dictionary<int, LiquidType>
```
Ideally maybe bouncer area?

hakusaro reviewed (approved):
Minor comments and suggestions, mostly about placement. No harm in not changing things in the spirit of getting them fixed though. :) :green_heart: thank you for working on this! :sparkles:

Olink committed:
Apply suggestions from code review

Co-authored-by: Lucas Nicodemus <shank@shanked.me>

Olink committed:
Move the projectile refresh to Bouncer.
Bouncer now has an OnSecondUpdate().
Move projectile created tracking to GetDataHandler from Bouncer.

Olink committed:
Add config option for the bomb radius.

Olink committed:
Add fluid rockets.

hakusaro commented on CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ This is the rolling changelog for TShock for Terraria. Use past tense when addin
 * Update for OTAPI 2.0.0.36 and Terraria 1.4.0.4. (@hakusaro, @Patrikkk, @DeathCradle)
 * Fixed /wind command. (@AxeelAnder)
 * Fixed NPC buff bouncer. (@AxeelAnder)
+* Fixed lava, wet, and honey bombs. (@Olink)
```
```suggestion
* Fixed lava, wet, honey bombs; and lava, wet, and honey rockets. (@Olink)
```

Olink committed:
Update CHANGELOG.md

Co-authored-by: Lucas Nicodemus <shank@shanked.me>

Olink committed:
Track projectile updates being sent by the server, instead of received by the server.

Olink committed:
Update changelog.

hakusaro commented on TShockAPI/ConfigFile.cs:
```diff
@@ -527,6 +527,9 @@ public class ConfigFile
 		[Description("Whether or not the server should output debug level messages related to system operation.")]
 		public bool DebugLogs = false;
 
+		[Description("Determines the range in tiles that a bomb can affect tiles.")]
```
```suggestion
		[Description("Determines the range in tiles that a bomb can affect tiles from detonation point.")]
```

hakusaro reviewed (approved):
Okay, looks good!

hakusaro committed:
Update config description for bomb range
